docs/specifications/tests/DNSSEC-TP/dnssec17.md

```diff
@@ -0,0 +1,213 @@
+8. If *DS Records* is empty, terminate this test case.
+
+9.  Retrieve all name server IP addresses for the *Child Zone* using
+    [Get-Del-NS-IPs] and [Get-Zone-NS-IPs] ("NS IP").
+
+10. Repeat the following steps for each name server IP address in 
+    *NS IP*:
+
+    1. Send the CDS query over UDP to the name server IP address.
+       1. If no DNS response is returned, then go to next name server 
+          IP.
+       2. Else, if AA bit is not set in the DNS response, then go to
+          next name server IP.
+       3. Else, if the RCODE in the DNS response is not *NOERROR*, then
+          go to next name server IP.
+       4. Else, if the DNS response contains at least one CDS record
+          in the answer section, then add the name server IP and the
+          CDS RRset to the *CDS RRsets* set. Also include any assciated
+          RRSIG records.
+    2. Send the CDNSKEY query over UDP to the name server IP address.
+       1. If no DNS response is returned, then go to next name server
+          IP.
+       2. Else, if AA bit is not set in the DNS response, then go to
+          next name server IP.
+       3. Else, if the RCODE in the DNS response is not *NOERROR*, then
+          go to next name server IP.
+       4. Else, if the DNS response contains at least one CDNSKEY
+          record in the answer section, then add the name server IP and
+          the CDNSKEY RRset from the answer section to the 
+          *CDNSKEY RRsets* set. Also include any assciated RRSIG records.
+    3. Send the DNSKEY query over UDP to the name server IP address.
+       1. If no DNS response is returned, then go to next name server
+          IP.
+       2. Else, if AA bit is not set in the DNS response, then go to
+          next name server IP.
+       3. Else, if the RCODE in the DNS response is not *NOERROR*, then
+          go to next name server IP.
+       4. Else, if the DNS response contains at least one DNSKEY
+          record in the answer section, then add the name server IP and
+          the CDNSKEY RRset from the answer section to the 
+          *DNSKEY RRsets* set.
+    4. Go to next name server IP.
+
+11. If neither of the *CDS RRsets* and *CDNSKEY RRsets* sets,
+    respectively, has any RRset then terminate this test case.
+
+12. If the *DNSKEY RRsets* is empty, then terminate this test case.
+
+13. For each name server IP in the *CDS RRsets* set do:
+
+    1. Extract the RRSIG records for the CDS RRset.
+    2. Extract the DNSKEY from the *DNSKEY RRsets* for the same name
+       server IP.
+    3. For each DS record in *DS Records* do:
+       1. If the DS record does not point at a DNSKEY record then add
+          name server IP and DS key tag to the 
+          *DS No Match CDS RRSIG*.
+       2. Else, if the DNSKEY that the DS record points at does not
+          match any RRSIG for CDS RRset then add name server IP and
+          DS key tag to the *DS No Match CDS RRSIG*.
+    4. Go to next name server IP address.
+
```

Comment:
This test will create false negatives.
Only one of the DS records has to match the RRSIG, not all.
Same applies below for CDNSKEY.

Reply:
> A DS RR SHOULD point to a DNSKEY RR that is present in the child's
> apex DNSKEY RRset, and the child's apex DNSKEY RRset SHOULD be
> signed by the corresponding private key."

https://tools.ietf.org/html/rfc4035#section-2.4

Reply:
Your code tests if all DS records point to an RRSIG.
RFC says "A DS" needs to point to the DNSKEY, sign the dnskey set and the cds set.
Your need to test backwards:
For all RRSIG of the CDS set check that there is an RRSIG with the same key on the dnskey set (from the same server ip) and check that there is a DS that points to the same key.

Reply:
My interpretation is that the CDS RRset should be signed by the private key of every DNSKEY that any DS points to.

Reply:
This test case doesn't test 4035, but https://tools.ietf.org/html/rfc7344#4.1
BTW, I haven't seen a test for continuity.

Reply:
> This test case doesn't test 4035, but https://tools.ietf.org/html/rfc7344#4.1

True, and "with a key" could be interpreted as "with at least one key". I will rework.

> BTW, I haven't seen a test for continuity.

That is part of the proposal for DNSSEC16 and DNSSEC17 in #939.